scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -473,18 +498,101 @@ static CoGbkResultSchema schemaOf(List<BucketedInput<?, ?>> sources) {
       return sourceMetadata;
     }
 
-    List<ResourceId> listFilesForBucket(int bucketId, int targetParallelism) {
-      return mapBucketFiles(bucketId, targetParallelism, Function.identity());
+    long getOrSampleByteSize(int bucketId, int targetParallelism) {
+      final AtomicLong totalNumFiles = new AtomicLong(0L);
+      flatmapBucketShardIds(
```

Comment:
This can simply be `flatmapBucketShardIds(..., (bucketShardId, metadata) -> 1).sum()`?

Reply:
Alternatively, `mapBucketShardIds` into `List<ResourceId>`, do `Collections.shuffle(list)`, and `take(SAMPLING_NUM_FILES_THRESHOLD`? That'll also get rid of the random fraction trick later.

And you can probably keep `mapBucketShardIds` as is.

Reply:
hmm, my idea here was to try to avoid materializing a list of (potentially) millions of `ResourceId`s (plus I think the GCS ResourceId/Path classes have some nontrivial initialization logic?), so the list is only ever <= 1000 ResourceIds.